### PR TITLE
feat: synchronize value for synchronized ranges

### PR DESCRIPTION
**Pull request by nathanielc**
This change updates the Recon protocol to check for any missing values when a synchronized range of keys is discovered. This way as nodes are synchronized they ensure they also have all values for their known keys.

TODO 
- [ ] The recon split changes need to merge first

### CODE REVIEW

File: recon/src/recon/sqlitestore.rs [hunk 1]
```diff
@@ -396,6 +413,31 @@ where
             .await?;
         Ok(row.map(|row| row.get(0)))
     }
+
+    #[instrument(skip(self))]
+    async fn keys_with_missing_values(
+        &mut self,
+        range: RangeOpen<Self::Key>,
+    ) -> Result<Vec<Self::Key>> {
+        let query = sqlx::query(
```
Comment (dav1do) [inline recon/src/recon/sqlitestore.rs:422]: when we need to check if the range is valid? In `hash_range`, we check if left >= right and return 0 but in other range related queries we don't.

File: recon/src/recon/tests.rs [hunk 2]
```diff
@@ -1427,6 +1479,80 @@ async fn two_in_sync() {
...
+            dog: [a: A, b: B, c: C]
+        -> range_req({𝚨 h(a, b, c)#3 𝛀 })
+            cat: [a: ∅, b: ∅, c: ∅]
+        <- range_resp({𝚨 h(a, b, c)#3 𝛀 })
+            dog: [a: A, b: B, c: C]
+        -> value_req(a)
+            cat: [a: ∅, b: ∅, c: ∅]
+        -> value_req(b)
+            cat: [a: ∅, b: ∅, c: ∅]
+        <- value_resp(a: A)
```
[the remark this answers is not in the document]
Comment (dav1do) [reply recon/src/recon/tests.rs:1536]: Along those lines.. in #243, I'm seeing some ordering changes to req/resp values cause failures and was hoping to get some help understanding what should happen and if it's a problem.

**Review by dav1do — COMMENTED** [review]
Looks pretty good. Has conflicts to resolve in order to run the checks. My open PR #243 will conflict but it shouldn't be too bad for me to resolve them if this goes in first.

File: recon/src/recon/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dav1do) [reply recon/src/recon/tests.rs:1538]: I regenerated the expect tests for my PR and they seem to be consistent in the that order for now. Will cause some more conflicts for you merging back in.. hopefully it's not bad. It should be merged shortly.

File: recon/src/recon.rs [hunk 3]
```diff
@@ -258,6 +258,10 @@ where
 
         Ok(new)
     }
+    /// Report all keys in the range that are missing a value
+    pub async fn keys_with_missing_values(&mut self, range: RangeOpen<K>) -> Result<Vec<K>> {
```
Comment (dav1do) [inline recon/src/recon.rs:262]: out of curiosity, why implement `range_with_values` and `keys_with_missing_values` rather than just `range_with_opt_values` and figure both out in one a level above storage? Not sure if it's ergonomics, the list of NULLs being big, or something I'm missing.

File: core/src/event_id.rs [hunk 4]
```diff
@@ -96,49 +122,92 @@ impl EventId {
...
+        let controller = &remainder[8..8 + 8];
+        // StreamID [u8; 4]
+        let stream_id = &remainder[8 + 8..8 + 8 + 4];
 
-        // strip separator [u8; 8] controller [u8; 8] StreamID [u8; 4]
         remainder = &remainder[(8 + 8 + 4)..];
 
         // height cbor unsigned integer
-        if remainder[0] <= 23 {
+        let (height, cid) = if remainder[0] <= 23 {
```
Comment (dav1do) [inline core/src/event_id.rs:137]: It might be worth extracting this to a small function, or at least adding a comment about why we're using big and little endian bytes. I had to look up how CBOR integer encoding is implemented to understand this.

**Review by dav1do — APPROVED** [review]
Looks good! A couple comments/questions, but I don't have any concerns that need to be addressed before merging.

File: core/src/event_id.rs (hunk 4 above)
Comment (nathanielc) [reply core/src/event_id.rs:137]: I had too read the spec as well to update the code. I'll add a comment and refactor into a function.

File: recon/src/recon.rs (hunk 3 above)
Comment (nathanielc) [reply recon/src/recon.rs:262]: In the case of keys_with_missing_values I don't want the values. So if I had to read all key/values in a range just to discard the values that actually existed that would be wasteful disk io.

File: recon/src/recon/sqlitestore.rs [hunk 5]
```diff
@@ -91,45 +97,64 @@ where
         // we insert the value first as it's possible we already have the key and can skip that step
         // as it happens in a transaction, we'll roll back the value insert if the key insert fails and try again
         if let Some(val) = item.value {
-            if self.insert_value_int(item.key, val, conn).await? {
+            // Update the value_retrieved flag, and report if the key already exists.
+            let key_exists = self.update_value_retrieved_int(item.key, conn).await?;
```
[the remark this answers is not in the document]
Comment (smrz2001) [reply recon/src/recon/sqlitestore.rs:101]: Yeah, I was just going to suggest combining the update/insert then saw your comment on refreshing, @dav1do 😄

**Review by smrz2001 — APPROVED** [review]
LGTM! Just one question about determinism in the tests.

File: recon/src/recon/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathanielc) [reply recon/src/recon/tests.rs:1538]: So its not perfectly deterministic but it does seem to be stable for a given implementation. Meaning I do not expect the tests to be flaky but changes in the performance of the code might change the order.

The test driver explicitly uses a single thread to drive the message exchange, but its still possible because of the async nature of the code that messages can be delivered in a different order. 

If we discover that the tests are somehow flaky we can explore ways to improve them. For example we could remove the assertions about state from each message and then allow for messages to be out of order using a new syntax. However I'd prefer to avoid that complexity if possible.

TL;DR we might have to update the expectations when we change to code but the tests should be stable.

[the remark this answers is not in the document]
Comment (smrz2001) [reply recon/src/recon/tests.rs:1538]: Sounds great, thanks.
